**Привет, Андрей!**

Меня зовут Анастасия Ефременко, я буду проверять твой проект. Предлагаю тебе общаться на "ты", но если тебе это некомфортно, дай знать и мы перейдем на "Вы". Цель этого ревью: помочь тебе довести проект до совершенства. Так, как это хотел бы видеть работодатель. 

Прошу тебя не удалять мои комменатрии пока проект не будет принят, чтобы облегчить нам процес работы.

Свои комменатрии я выделяю разными цветами, и вот что это значит:

<div class="alert alert-danger">
<b>Осторожно!</b> 
Красным цветом выделены недочеты, на которые обязательно нужно обратить внимание. Их следует исправить в твоей работе.</div>

<div class="alert alert-warning">
<b>Рекомендация:</b> 
Желтым цветом я оставляю тебе советы, которые стоит принять во внимание. Если есть желание, можешь исправить их. </div>

<div class="alert alert-success">
<b>Good job! </b> Ну а зеленым цветом будет помечена обратная связь по твоему проекту.</div>

Предлагаю тебе работать в тандеме, поэтому ответы на мои комментарии, возникающие по ходу работы вопросы можешь выделять синим цветом.
<div class="alert alert-info">
<b>Комментарий студента:</b> выделяется синим цветом.</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success">
<b>Комментарий ревьюера: </b> Отлично, что в начале проекта есть его краткое описание и содержание. Это поможет более точно расставлять акценты в выводах. Также здорово, что ты добавил описание данных.   </div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
print('Дубликатов в данных:', data.duplicated().sum()) 
display(data)
data.info()

Дубликатов в данных: 54


children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-warning">
<b>Комментарий ревьюера: </b>Для того, чтобы информация у дублях вывелась в одной строке, лучше было бы использовать `print()`, а не `display()`. `display()` используют, как правило, только для того, чтобы вывести датафрейм в виде таблицы. </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Исправил display() на print() здесь и далее. 
    
Я плохо владею юпитером, многое делаю методом тыка.
Различия между display() и print() по сути не знаю, а времени на изучение юпитера пока не было.
</div>




<div class="alert alert-success">
<b>Комментарий ревьюера v2: </b>

`display()` используют для того, чтобы вывести датафрйм в виде таблицы. Во всех остальных случаях используют `print()`. Но они не нужны для последней строки ячейки, так как там Jupyter все сделает автоматически.
  
Вот и вся разница.
    
Jupyter еще изучишь в ходе обучения, можешь подглядывать за всякими лайфхаками, которые вставляют в свои комментарии ревьюеры.
</div>

Описание данных

•	children — количество детей в семье

•	days_employed — общий трудовой стаж в днях

•	dob_years — возраст клиента в годах

•	education — уровень образования клиента

•	education_id — идентификатор уровня образования

•	family_status — семейное положение

•	family_status_id — идентификатор семейного положения

•	gender — пол клиента

•	income_type — тип занятости

•	debt — имел ли задолженность по возврату кредитов

•	total_income — ежемесячный доход

•	purpose — цель получения кредита

••••• ••••• ••••• Выводы из данных таблицы ••••• ••••• •••••

   • В таблице 21525 строк. В двух столбцах одинаковое количество строк с пропущенными значениями: days_employed и total_income, т.е. общий трудовой стаж в днях и ежемесячный доход. Можно предположить, что пропуски неслучайны, т.к. отсутствующие данные взаимосвязаны.  Причина пропусков может быть различной: сокрытие данных клиентом, невозможностью предоставить подтверждающих "трудовой стаж" и "ежемесячный заработок" документов, что возможно в случае, например, неофициальной работы, утерей подтверждающих документов, работой за границей, отбыванием наказания в исправительных учреждениях и пр. Заменив пропущенные значения нулями, я изменю данные с неизвестного на нулевой доход, что принципиально различается. Также неправомерно заменять пропущенные данные медианными значениями, средними и любыми другими. Если данных нет, нет признаков и достоверных способов исправления данных, то все замены ни на чём не основанная чистая фантазия. Аналогично со столбцом days_employed. Однако, шаблон проекта обязывает обрабатывать ячейки с пропущенными значениями. Для заполнения пропущенных значений рассчитываю медианное значение трудового стажа в зависимости от возраста клиента (пользуясь данными всей таблицы). Пропущенные данные в столбце ежемесячного дохода заменяю медианными значениями ежемесячного дохода в зависимости от уровня образования клиента. Метод спорный, неоднозначный.

   
   • Столбец days_employed содержит отрицательные числовые значения (отрицательное количество дней), что не имеет физического смысла. Это значит, что нужно перевести данные в абсолютное значение, оставить значение по модулю.
   
   • Столбец days_employed уже в строке 4 содержит значение (по модулю): 340266, что в пересчёте равно 932 годам, что не имеет уже биологического смысла. Очевидно, это искажённые/ошибочные данные, которые нужно исправлять. Если исходить из того, что человек мог начать работать подростком, а на момент исследования находится в глубокой старости, то максимальный трудовой стаж можно ограничить значением в 25000 дней (68,5 лет).          
      
   • Столбец days_employed содержит данные типа float64. Значение общего трудового стажа в днях с точностью до 6-ти знаков после запятой - в такой точности нет необходимости. Тип данных нужно перевести в тип int64.
   
   • Столбец total_income/ежемесячный доход содержит данные типа float64. С учётом величины значения,  точность до 6-ти знаков после запятой не нужна. Тип данных нужно перевести в тип int64.

   • Таблица содержит дубликаты. На этом этапе выявлено 54 явных дубликата. На следующих этапах нужно свести все буквенные записи к нижнему регистру (для нахождения неявных дубликатов), дубликаты удалить.
   
   • Замена типа данных проводится методом astype(). Потому что так легче, быстрее, а другие способы на ум не пришли, да и не знаю я их.

<div class="alert alert-success">
<b>Комментарий ревьюера: </b> Классное подробное знакомство с данными, молодец!  </div>

## Шаг 2. Предобработка данных

### Обработка пропусков и замена типа данных

In [2]:
# для подстраховки от искажений перевожу в абсолютное значение значения столбца dob_years
data['dob_years'] = abs(data['dob_years'])

# шаблон задачи настаивает на замене пропущенных значений. ввожу переменную median_from_age, 
# рассчитывающую медианное значение трудового стажа в зависимости от возраста клиента.
# ставлю полученное из расчётов median_from_age в пустую ячейку столбца 'days_employed' 
# клиентам того же возраста
for age in data['dob_years'].unique():
    median = data.loc[data['dob_years'] == age, 'days_employed'].median()
    data.loc[(data['days_employed'].isna()) & (data['dob_years'] == age), \
    'days_employed'] = median

# аналогичную замену делаю для пропущенных значений в столбце total_income, отталкиваясь от 
# уровня образования клиента
for education_id in data['education_id'].unique():
    median_from_ed = data.loc[data['education_id'] == education_id, 'total_income'].median()
    data.loc[(data['total_income'].isna()) & (data['education_id'] == education_id), \
    'total_income'] = median_from_ed  
    
# перевожу тип даных из float в int
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')    
    
# перевод значений в абсолютное значение, избавление от отрицательных значений
data['days_employed'] = abs(data['days_employed'])

# количество строк с неестественно большим/ошибочным значением дней трудового стажа, исходя из 
# максимума в 25000 дней
print('Строк со слишком большим трудовым стажем:',\
len(data.loc[data.loc[:,'days_employed'] > 25000])) 

# сведение всех символовв данных столбца education и family_status к нижнему регистру
data['education'] = data['education'].str.lower()
data['family_status'] = data['family_status'].str.lower()

# контроль выполнения кода
display(data.head(10))
data.info()

Строк со слишком большим трудовым стажем: 3853


children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

Строк с неестественно большим/ошибочным значением дней трудового стажа, исходя из максимума в 25000 дней, оказалось значительным - 3853, т.е., почти 18% всех строк. При этом, в таблице присутствуют слишком малые значения дней трудового стажа - единицы сотен дней у взрослого человека (например, строка 7 - 152 дня у 50-ти летнего мужчины).

Варианты действий с такими строками и оценка вариантов:

1. Оставить строки в неизменном виде, предполагая, что для целей исследования трудовой стаж не имеет значения, а достоверного способа приведения данных в нормальный вид нет. - РЕШАЮ ПОСТУПАТЬ ТАК.

2. Удалить строки, предполагая, что ошибочные данные исказят результаты исследования. - Количество строк слишком велико, чтобы жертвовать таким объёмом информации, без чётких данных, что неверное значение трудового стажа искажает результаты исследования. 

3. Заменить значение максимальным, минимальным, средним или медианным значением. - На перечисленные действия нет полномочий, т.к. такие действия не подкреплены никакими предпосылками.

4. Ввести поправочный коэффициент, поделив неверные данные на этот коэффициент, уменьшив значение до разумного. Этот вариант исходит из предположения, что сместилась запятая, или по каким-то причинам все данные умножились на одно значение, или прибавилось некое значение. - Нет информации о причине искажения данных, нет поправочного коэффициента, нет возможности привести данные к нормальным значениям. Плюс к перечисленному, данные искажены неравномерно и неодинаково, что не позволяет подойти к обработке данных единым строго определённым образом.


<div class="alert alert-success">
<b>Комментарий ревьюера: </b> Отличная подробная аргументация принятых решений, молодец!

Отличная идея с заполнением пропусков медианой с предварительной группиировкой. Для дохода логичнее было бы использовать не образование, а тип занятости. Например, доход пенсионера явно ниже, чем доход предпринимателя.

Также верная работа над неявными дублями из-за разного регистра.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера: </b> 

Здорово, что ты заметил выбросы в стаже. Как вариант, можно предположить, что эти значения просто указаны в часах, а не днях, и поделить их на соответвующее значение.
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

В том то и дело, что где-то в стаже огромные значения, а где-то очень маленькие - полгода за всю жизнь у взрослого человека. Даже если у таких заменять дни месяцами (умножать на 30), то получиться менее 15-ти лет за всю жизнь - мало как-то. И эти значения разбросаны случайным образом.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера v2: </b>

На самом деле, не случайным. Эти выбросы есть только у пенсионеров и безработных. Причем у них все значения такие. Или же можно заметить, что такие огромные только положительные значения, отрицательные же вполне адекватные. Так что можно было бы просто разделить все изначально положителньые значения на 24 и привести их к часам.
</div>

### Предпросмотр типа данных для последующей категоризации

In [3]:
# оценка количества типов занятости и целей кредита для последующей категоризации
display(data['income_type'].value_counts())
display(len(data['income_type'].value_counts()))
display(data['purpose'].value_counts())
display(len(data['purpose'].value_counts()))

# вывод данных для оценки наличия или отсутвтстия ошибок в данных
display('Значение value_counts для столбца debt', data['debt'].value_counts())
display('Значение value_counts для столбца gender', data['gender'].value_counts())
display('Значение value_counts для столбца education_id', data['education_id'].value_counts())
display('Значение value_counts для столбца family_status_id', data['family_status_id'].value_counts())
display('Значение value_counts для столбца children', data['children'].value_counts())

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

8

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

38

'Значение value_counts для столбца debt'

0    19784
1     1741
Name: debt, dtype: int64

'Значение value_counts для столбца gender'

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

'Значение value_counts для столбца education_id'

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

'Значение value_counts для столбца family_status_id'

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

'Значение value_counts для столбца children'

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

**Вывод**

В столбце income_type (тип занятости) представлено 8 типов, в достаточной мере различных. Нужды в объединении типов для уменьшения количества типов занятости нет.

В столбце purpose (цель кредита) представлено 38 целей, многие из которых можно объединить, тем самым сильно сократив количество целей.

Столбец debt (данные по задолженности) имеет только 2 значения - 0 и 1. Отрицательных и прочих ошибочных значений не имеет.

Столбец gender (пол) имеет 3 значения - F, M и только в одной строке значение XNA. Одна строка с неопределённым полом клиента имеет слишком малый информационный вес. Оставляю строку без изменений.

Столбец education_id (категоризированные данные об образовании) имеет 5 значений - от 0 до 4 без пропусков. Отрицательных и прочих ошибочных значений не имеет.

Столбец family_status_id (категоризированные данные о семейном статусе) имеет 5 значений - от 0 до 4 без пропусков. Отрицательных и прочих ошибочных значений не имеет.

Столбец children кроме нормальных значенийот 0 до 5 без пропусков и с нормальным распределением, имеет аномальные значения: минус 1; 20 детей. Аномальные значения встречаются не в единичных случаях, а в общей сумме в более, чем в 100 строках таблицы. Т.к., по условию проекта, аномальные значения нужно обрабатывать по своему усмотрению, и учитывая нормальное распределение количества детей, делаю следующие предположения: 
* значение "минус 1" - ошибочно записанное значение "1";
* значение "20" - ошибочно записанное значение "0".

Предположение о том, что "20" - это неверно записанный "0", а не неверно записанная "2" исходит из того, что, на клавиатуре клавишт "0" и "2" расположены рядом, и судя по распределению количества детей, количество "0" детей встречается в 7 раз чаще, чем количество детей "2"; значит, более вероятно значение "0" детей. Далее делаю соответствующую замену.


<div class="alert alert-success">
<b>Комментарий ревьюера: </b> Отличная подробная аргументация принятых решений, молодец!

Идеальное подробное описание данных и аругментация принятых решений.
</div>

### Обработка дубликатов

In [4]:
# замена аномальных значений количества детей на нормальное.
# контроль выполнения замены
data.loc[data['children'] == 20, 'children'] = 0
data.loc[data['children'] == -1, 'children'] = 1
display('Значение value_counts для столбца children', data['children'].value_counts())

'Значение value_counts для столбца children'

0    14225
1     4865
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

In [5]:
# удаление дубликатов и контроль выполнения кода
print('До обработки дубликатов:', data.duplicated().sum()) 
data = data.drop_duplicates().reset_index(drop=True)
print('После обработки дубликатов:', data.duplicated().sum()) 

До обработки дубликатов: 71
После обработки дубликатов: 0


In [6]:
# лемматизация и подсчёт лемм слов в столбце 'purpose' для дальнейшей категоризации
from pymystem3 import Mystem
m = Mystem() 
h = m.lemmatize(' '.join(data['purpose']))

from collections import Counter
print(Counter(h)) 

Counter({' ': 55023, 'недвижимость': 6351, 'покупка': 5897, 'жилье': 4460, 'автомобиль': 4306, 'образование': 4013, 'с': 2918, 'операция': 2604, 'свадьба': 2324, 'свой': 2230, 'на': 2222, 'строительство': 1878, 'высокий': 1374, 'получение': 1314, 'коммерческий': 1311, 'для': 1289, 'жилой': 1230, 'сделка': 941, 'дополнительный': 906, 'заниматься': 904, 'подержать': 853, 'проведение': 768, 'сыграть': 765, 'сдача': 651, 'семья': 638, 'собственный': 635, 'со': 627, 'ремонт': 607, 'приобретение': 461, 'профильный': 436, 'подержанный': 111, '\n': 1})


<div class="alert alert-warning">
<b>Комментарий ревьюера: </b> 

На будущее рекомендую разбивать на ячейки код, который относится к разным столбцам, по смыслу.
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Понял. Исправил.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2: </b>

Так проект стал намного читабельнее, спасибо! Тебе это тоже пригодится, когда через какое-то время придется вернуться к проекту. По ячейкам быстрее сообразить и вспомнить, о чем шла речь.
</div>

**Вывод**

До обработки дубликатов: было 71. После обработки стало ноль дубликатов. За неимением уникальных идентификаторов клиентов, поиск дубликатов ведётся по всем столбцам таблицы.

Замена аномальных значений количества детей не изменила нормального распределения количества.

Дубликаты удаляю методом drop_duplicates(). Удадяю этим методом, потому что других методов не знаю. Без уникального идентификационного номера неправомочно утверждать, что одинаковые строки - дубликаты, тем более, с учётом такого большого количества пропусков. Поэтому причина появления дубликатов - необходимость включения дубликатов в учебный проект. Ну и ошибка оператора, вероятно, самая распространённая причина появления дубликатов.

Лемматизация совместно с методом value_counts продемонстрировала небольшой набор целей кредита. Вручную, руководствуясь логикой, я отобрал самые популярные слова. 


<div class="alert alert-success">
<b>Комментарий ревьюера: </b> Полностью согласна с принятыми решениями по выбросам в количестве детей.
    
Также количество лемм посчитано верно, это позволяет выделить исчерпывающее количество категорий.
</div>

### Лемматизация

In [7]:
# присвоение целям кредита сокращённой записи на основе самых популярных лемм 
# из прежних формулировок целей кредита
      
from pymystem3 import Mystem
m = Mystem() 

def purpose_func(purpose_lemmas):
    purposes = ['жилье', 'недвижимость', 'автомобиль', 'образование', 'свадьба']
    purpose_lemmas = m.lemmatize(purpose_lemmas['purpose'])
    for purpose in purposes:
        if purpose in purpose_lemmas:  
            return purpose
data['purpose'] = data.apply(purpose_func, axis=1)

# из-за практической идентичности понятий 'жилье' и 'недвижимость' переименовываю цели, 
# таким образом объединяя эти одинаковые цели в одну
data.loc[data['purpose'] == 'жилье', 'purpose'] = 'недвижимость' 

# подсчёт количества целей=лемм, контроль выполнения последних действий по объединению целей
from collections import Counter
print(Counter(data.apply(purpose_func, axis=1))) 

Counter({'недвижимость': 10811, 'автомобиль': 4306, 'образование': 4013, 'свадьба': 2324})


**Вывод**

Я выделил 5 самых часто встречаемых слов в формулировках целей кредита, и свёл все цели кредита к этим четырём целям.

Шаблон решения проектной работы обязывает использование лемматизации, что, по моему мнению, в этой задаче не нужно, т.к. формулировок целей немного - 38, разбивка в ручную, и присвоение в цикле в этом случае даст разбивку на подцели гораздо точнее. Например, сумма кредита на ремонт жилья скорее всего меншье, чем на строительство или покупку жилья, а высшее образование стоит дороже и сам процесс обучения длится дольше по сравнению с дополнительным образованием. Тоже самое с подержанными автомобилями и просто автомобилями, жильём для собственного использования или для коммерческих целей.

<div class="alert alert-success">
<b>Комментарий ревьюера: </b> 
    
В данной задаче, действительно, не так много значений. Цель здесь - научить делать что-то +- универсальное, чтобы в задаче, где больше уникальных значений, студенты не растерялись.
    
Ты подметил интересный детали, в реальном проекте можно было бы попросить чуть больше данных (с суммой кредита) и все это проверить.
    
Также я полностью согласна с тем, что ты объединил категорию недвижимости и жилья.
</div>


### Категоризация данных

In [8]:
# просмотр уровня дохода клиентуры для послеждующей категоризации по уровню дохода
print(data['total_income'].min())
print(data['total_income'].max())
print(data['total_income'].mean())
print(data['total_income'].median())

# функция, придающая статус клиенту в соответствии с количеством у него детей
def kindergarten(children):
    if children == 0:
        return 'бездетный'
    elif 0 < children < 3:
        return 'обычнодетный'
    else:
        return 'многодетный'
    
data['kindergarten'] = data['children'].apply(kindergarten) 

data['richness'] = pd.qcut(data.total_income, 4, labels=["низкий доход", "ниже среднего", "выше среднего", "высокий"])

display(data.head(10))


20667
2265604
165412.95884217395
141147.0


children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

        purpose  kindergarten       richness  
0  недвижимость  обычнодетный        высокий  
1    автомобиль  обычнодетный  ниже среднего  
2  недвижимость     бездетный  выше среднего  
3   образование   многодетный        высокий  
4       свадьба     бездетный  выше среднего  
5  недвижимость     бездетный        высокий  
6  недвижимость     бездетный        высокий  
7   образование     бездетный  ниже среднего  
8       свадьба  обычнодетный   низкий доход  
9  недвижимость     бездетный  выше среднего

**Вывод**

<div class="alert alert-success">
<b>Комментарий ревьюера: </b> Хорошая идея категоризации по количеству детей.
</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера: </b> 
Когда есть возможность самостоятельно разделить клиентов на категории нужно стремиться к одинаковому размеру и равномерному распределению для более достоверных результатов. Для этих целей рекомендую использовать процентили. Например, с помощью 33 и 66 процентиля можно разделить выборку на 3 равные группы. А с помощью квартилей (25, 50 и 75 процентиль) - на 4. Квартили ты можешь увидеть с помощью метода describe().

Квартили означают следующее:

* 25 процентиль - 25% выборки имеют доход ниже этого значения;
* 50 процентиль - половина выборки имеют доход, ниже этого значения;
* 75 процентиль - 25% выборки имеют доход выше этого значения.
    
Чтобы разделить выборку на 4 равные группы, есть специальный метод, который делит выборку по квантилям на заданное количество групп:
    
    "
    pd.qcut(data.total_income, 4, labels=["низкий доход", "ниже среднего", "выше среднего", "высокий"])
    "
Либо ты можешь сделать это от руки, используя числа из описательной статистики (`.describe()`)  </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

А вот здесь я категорически и абсолютно не согласен. 
    
Почему нужно стремиться к одинаковому размеру? 
    
Пример: выборка из 100 человек, из них 90 человек имеют доход: 20-30 тысяч, 5 человек: 500 тысяч, 5 человек: 2 млн. И каков смысл разбивать такую выборку на одинаковые по численности группы?
    
Метод .describe() показывает, как идёт объединение (я посмотрел). Получается, что доход от около 200 тысяч до 2 млн.200 тысяч в одной группе "высокий" - в этом нет никакого смысла.
    
Я вижу, что квартили / процентили пишут в замечания к проектам часто (а может и всем), видимо, это такой образовательный приём, и без этого проект не будет принят, я подозреваю. Соответственно, я вынужден принять это, как в совковой школе: учитель всегда прав. (Это не в ваш адрес, а в адрес создателей курса).
    
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2: </b>

Да нет, ты не вынужден делать так, как сказал ревьюер. Если ты на 100% уверен в решени и аргументы ревьюера тебя не переубедили, то ты можешь сделать так, как считаешь правильным. 
    
Категоризация по количественным данным - это тонкий лед :) Если у банка уже есть критерии разделения на категории, то мы безусловно принимаем их. Также в основу ты мог бы положить какую-нибудь категоризацию признанного источника. Тут важно не впадать в самодурство, понимаешь, чтобы были какие-то осязаемые критерии выделения.
    
А в том, чтобы делить выборку на примерно равные группы есть важное преимущество - мы можем использовать разные статистические методы, что возможно только при равных выборках. Это будет изучаться далее.
</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [9]:
# создание таблицы со столбцами 'kindergarten' и 'debt'
children_vs_debt = data.loc[:, ['kindergarten', 'debt']]

# вывод общего числа клиентов, имевших задолженность по возврату кредитов
total_debt_1 = len(children_vs_debt.loc[children_vs_debt['debt'] == 1])
print('Общее число клиентов, имевших задолженность по возврату кредитов:', total_debt_1)

# расчёт и вывод процента бездетных, имевших задолженность по возврату кредитов
debt_nonchilds = len(children_vs_debt.loc[(children_vs_debt['debt'] == 1)\
& (children_vs_debt['kindergarten'] == 'бездетный')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, бездетных клиентов: \
{debt_nonchilds:.2%}')

# расчёт и вывод процента обычнодетных, имевших задолженность по возврату кредитов
debt_ordinary_childs = len(children_vs_debt.loc[(children_vs_debt['debt'] == 1)\
& (children_vs_debt['kindergarten'] == 'обычнодетный')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, обычнодетных (имеющих 1 или 2 детей) клиентов: \
{debt_ordinary_childs:.2%}')

# расчёт и вывод процента многодетных, имевших задолженность по возврату кредитов
debt_many_childs = len(children_vs_debt.loc[(children_vs_debt['debt'] == 1)\
& (children_vs_debt['kindergarten'] == 'многодетный')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, многодетных (имеющих от 3 до 5 детей) клиентов: \
{debt_many_childs:.2%}')

# контроль пропущенных строк с данными о задолженности
print('Строк, где debt и не ноль, и не единица:', \
len(children_vs_debt.loc[(children_vs_debt['debt'] != 1)
& (children_vs_debt['debt'] != 0)]))




Общее число клиентов, имевших задолженность по возврату кредитов: 1741
Из общего числа, имевших задолженность по возврату кредитов, бездетных клиентов: 61.52%
Из общего числа, имевших задолженность по возврату кредитов, обычнодетных (имеющих 1 или 2 детей) клиентов: 36.70%
Из общего числа, имевших задолженность по возврату кредитов, многодетных (имеющих от 3 до 5 детей) клиентов: 1.78%
Строк, где debt и не ноль, и не единица: 0


**Вывод**

Общее число клиентов, имевших задолженность по возврату кредитов: 1741 человек. Было показано, что строк в таблице, не попавших в это исследование - ноль.
Приняв 1741 человека за 100%, путём рассчётов показал, что из 1741 человека, имевших задолженность по возврату кредитов:
- бездетные клиенты составляют: 61.52%
- клиенты имеющие 1 или 2 детей, составляют: 36.70%
- многодетные клиенты, т.е., имеющие от 3 до 5 детей, составляют: 1.78%.

Отсюда можно сделать вывод, что зависимость между наличием детей и возвратом кредита в срок существует, и она такова:
чем больше детей у клиента (в рамках от нуля до пяти детей), тем выше вероятность возврата кредита в срок.


- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [10]:
# таблица - расшифровка статусов семейного положения
family_id_is = data[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)
display(family_id_is) 

# создание таблицы со столбцами 'family_status_id' и 'debt'   
family_vs_debt = data.loc[:, ['family_status_id', 'debt']]

# общее число клиентов, имевших задолженность по возврату кредитов берём из предыдущего блока 
# total_debt_1 = 1741

# расчёт и вывод процента женатых/замужних, имевших задолженность по возврату кредитов
debt_family_0 = len(family_vs_debt.loc[(family_vs_debt['debt'] == 1)\
& (family_vs_debt['family_status_id'] == 0)]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, женатых и замужних: \
{debt_family_0:.2%}')

# расчёт и вывод процента состоящих в гражданском браке, имевших задолженность по возврату кредитов
debt_family_1 = len(family_vs_debt.loc[(family_vs_debt['debt'] == 1)\
& (family_vs_debt['family_status_id'] == 1)]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов клиентов, состоящих в гражданском браке: \
{debt_family_1:.2%}')

# расчёт и вывод процента вдовцов и вдов, имевших задолженность по возврату кредитов
debt_family_2 = len(family_vs_debt.loc[(family_vs_debt['debt'] == 1)\
& (family_vs_debt['family_status_id'] == 2)]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, вдовцов и вдов: \
{debt_family_2:.2%}')

# расчёт и вывод процента разведённых, имевших задолженность по возврату кредитов
debt_family_3 = len(family_vs_debt.loc[(family_vs_debt['debt'] == 1)\
& (family_vs_debt['family_status_id'] == 3)]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, клиентов в разводе: \
{debt_family_3:.2%}')

# расчёт и вывод процента незамужних и неженатых, имевших задолженность по возврату кредитов
debt_family_4 = len(family_vs_debt.loc[(family_vs_debt['debt'] == 1)\
& (family_vs_debt['family_status_id'] == 4)]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, незамужних и неженатых клиентов: \
{debt_family_4:.2%}')


family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

Из общего числа, имевших задолженность по возврату кредитов, женатых и замужних: 53.48%
Из общего числа, имевших задолженность по возврату кредитов клиентов, состоящих в гражданском браке: 22.29%
Из общего числа, имевших задолженность по возврату кредитов, вдовцов и вдов: 3.62%
Из общего числа, имевших задолженность по возврату кредитов, клиентов в разводе: 4.88%
Из общего числа, имевших задолженность по возврату кредитов, незамужних и неженатых клиентов: 15.74%


**Вывод**

Общее число клиентов, имевших задолженность по возврату кредитов: 1741 человек. 

Приняв 1741 человека за 100%, путём рассчётов показал, что из 1741 человека, имевших задолженность по возврату кредитов:
- женатые и замужние: 53.48%
- состоящие в гражданском браке: 22.29%
- вдовцы и вдовы: 3.62%
- состоят в разводе: 4.88%
- незамужние и неженатые: 15.74%    
    
Отсюда можно сделать вывод, что зависимость между семейным положением и возвратом кредита в срок существует, и она такова: вероятность возврата кредита в срок выше у одиноких людей (вдовцы, вдовы, разведённые); вероятность невозврата кредита в срок выше у женатых и замужних людей.


- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [11]:
# создание таблицы со столбцами 'richness' и 'debt'   
richness_vs_debt = data.loc[:, ['richness', 'debt']]

# общее число клиентов, имевших задолженность по возврату кредитов берём из предыдущего блока 
# total_debt_1 = 1741

# расчёт 
debt_richness_0 = len(richness_vs_debt.loc[(richness_vs_debt['debt'] == 1)\
& (richness_vs_debt['richness'] == 'низкий доход')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, имеют низкий доход: \
{debt_richness_0:.2%}')

# расчёт 
debt_richness_1 = len(richness_vs_debt.loc[(richness_vs_debt['debt'] == 1)\
& (richness_vs_debt['richness'] == 'ниже среднего')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов имеют доход ниже среднего: \
{debt_richness_1:.2%}')

# расчёт 
debt_richness_2 = len(richness_vs_debt.loc[(richness_vs_debt['debt'] == 1)\
& (richness_vs_debt['richness'] == 'выше среднего')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов имеют доход выше среднего: \
{debt_richness_2:.2%}')

# расчёт 
debt_richness_3 = len(richness_vs_debt.loc[(richness_vs_debt['debt'] == 1)\
& (richness_vs_debt['richness'] == 'высокий')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, имеют высокий доход: \
{debt_richness_3:.2%}')


Из общего числа, имевших задолженность по возврату кредитов, имеют низкий доход: 24.53%
Из общего числа, имевших задолженность по возврату кредитов имеют доход ниже среднего: 27.17%
Из общего числа, имевших задолженность по возврату кредитов имеют доход выше среднего: 26.31%
Из общего числа, имевших задолженность по возврату кредитов, имеют высокий доход: 22.00%


**Вывод**

Общее число клиентов, имевших задолженность по возврату кредитов: 1741 человек.

Приняв 1741 человека за 100%, путём рассчётов показал, что из 1741 человека, имевших задолженность по возврату кредитов:

- малоимущих (имеющих заработок до 80 000 рублей в месяц): 9.99%
- со средним заработком (имеющих заработок от 80 000  до 180 000 рублей в месяц): 62.55%
- обеспеченных (имеющих заработок свыше 180 000 рублей в месяц): 27.46%

Отсюда можно сделать вывод, что зависимость между семейным положением и возвратом кредита в срок существует, и она такова: вероятность возврата кредита в срок выше у малоимущих людей (имеющих заработок до 80 000 рублей в месяц); максимальная (в рамках этого исследования) вероятность невозврата кредита в срок у людей со средним заработком (имеющих заработок от 80 000  до 180 000 рублей в месяц).

Увы, код с расчётом приведённых выше данных пришлось удалить. Получившиеся данные даже не хочу комментировать - они ни о чём не говорят, т.к. сам способ их получения не позволяет получить реперзентативные данные.


<div class="alert alert-info">
<h2> Комментарий студента</h2>

Разбив клиентов по уровню дохода на равные по численности группы, я получил совершенно нерепрезентативные данные - размазанные проценты, также примерно равно делящиеся. Это легко объяснить: разница в доходах между категориями "ниже среднего" и "выше среднего" несколько десятков тысяч рублей, а внутри категории "высокий доход" разница в доходах более, чем в 11 раз. 
    
Из общего числа, имевших задолженность по возврату кредитов, имеют низкий доход: 24.53%
    
Из общего числа, имевших задолженность по возврату кредитов имеют доход ниже среднего: 27.17%
    
Из общего числа, имевших задолженность по возврату кредитов имеют доход выше среднего: 26.31%
    
Из общего числа, имевших задолженность по возврату кредитов, имеют высокий доход: 22.00%
    
    
    По моему мнению, теперь это полная чушь, а не исследование.
    
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [12]:
# создание таблицы со столбцами 'purpose' и 'debt'   
purpose_vs_debt = data.loc[:, ['purpose', 'debt']]

# общее число клиентов, имевших задолженность по возврату кредитов берём из предыдущего блока 
# total_debt_1 = 1741

# расчёт и вывод процента малоимущих, имевших задолженность по возврату кредитов
debt_house = len(purpose_vs_debt.loc[(purpose_vs_debt['debt'] == 1)\
& (purpose_vs_debt['purpose'] == 'недвижимость')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, имеют цель кредита: недвижимость: \
{debt_house:.2%}')

# расчёт и вывод процента среднезарабатывающих, имевших задолженность по возврату кредитов
debt_auto = len(purpose_vs_debt.loc[(purpose_vs_debt['debt'] == 1)\
& (purpose_vs_debt['purpose'] == 'автомобиль')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов имеют цель кредита: автомобиль: \
{debt_auto:.2%}')

# расчёт и вывод процента обеспеченных, имевших задолженность по возврату кредитов
debt_education = len(purpose_vs_debt.loc[(purpose_vs_debt['debt'] == 1)\
& (purpose_vs_debt['purpose'] == 'образование')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, имеют цель кредита: получение образования: \
{debt_education:.2%}')

# расчёт и вывод процента обеспеченных, имевших задолженность по возврату кредитов
debt_wedding = len(purpose_vs_debt.loc[(purpose_vs_debt['debt'] == 1)\
& (purpose_vs_debt['purpose'] == 'свадьба')]) / total_debt_1
print(f'Из общего числа, имевших задолженность по возврату кредитов, имеют цель кредита: проведение свадьбы: \
{debt_wedding:.2%}')



Из общего числа, имевших задолженность по возврату кредитов, имеют цель кредита: недвижимость: 44.92%
Из общего числа, имевших задолженность по возврату кредитов имеют цель кредита: автомобиль: 23.15%
Из общего числа, имевших задолженность по возврату кредитов, имеют цель кредита: получение образования: 21.25%
Из общего числа, имевших задолженность по возврату кредитов, имеют цель кредита: проведение свадьбы: 10.68%


**Вывод**

Общее число клиентов, имевших задолженность по возврату кредитов: 1741 человек.

Приняв 1741 человека за 100%, путём рассчётов показал, что из 1741 человека, имевших задолженность по возврату кредитов:

- имеют цель кредита: недвижимость: 44.92%
- имеют цель кредита: автомобиль: 23.15%
- имеют цель кредита: получение образования: 21.25%
- имеют цель кредита: проведение свадьбы: 10.68% 

Отсюда можно сделать вывод, что цели кредита влияют на его возврат в срок таким образом: вероятность возврата кредита в срок выше у кредита, цель которого проведение свадебного торжества; максимальная (в рамках этого исследования) вероятность невозврата кредита в срок у кредита, взятого с целью операций с недвижимостью.


In [28]:
# данные по количеству детей
print('Бездетных:', len(data.loc[data['kindergarten'] == 'бездетный'])) 
print('Обычнодетных:', len(data.loc[data['kindergarten'] == 'обычнодетный'])) 
print('Многодетных:', len(data.loc[data['kindergarten'] == 'многодетный'])) 
print()
                                                                                                                  
# данные по уровню дохода
print('Низкий доход имеют:', len(data.loc[data['richness'] == 'низкий доход'])) 
print('Ниже среднего доход имеют:', len(data.loc[data['richness'] == 'ниже среднего'])) 
print('Выше среднего доход имеют:', len(data.loc[data['richness'] == 'выше среднего'])) 
print('Высокий доход имеют:', len(data.loc[data['richness'] == 'высокий'])) 
print()

# данные по семейному статусу
print('Семейное положение: женат/замужем', len(data.loc[data['family_status_id'] == 0])) 
print('Семейное положение: гражданский брак', len(data.loc[data['family_status_id'] == 1])) 
print('Семейное положение: вдова/вдовец', len(data.loc[data['family_status_id'] == 2])) 
print('Семейное положение: в разводе', len(data.loc[data['family_status_id'] == 3]))
print('Семейное положение: не женат/не замужем', len(data.loc[data['family_status_id'] == 4])) 
print()

# данные по целям кредита
print('Цель кредита: недвижимость', len(data.loc[data['purpose'] == 'недвижимость'])) 
print('Цель кредита: автомобиль', len(data.loc[data['purpose'] == 'автомобиль'])) 
print('Цель кредита: образование', len(data.loc[data['purpose'] == 'образование'])) 
print('Цель кредита: свадьба', len(data.loc[data['purpose'] == 'свадьба'])) 

Бездетных: 14167
Обычнодетных: 6907
Многодетных: 380

Низкий доход имеют: 5364
Ниже среднего доход имеют: 5363
Выше среднего доход имеют: 5363
Высокий доход имеют: 5364

Семейное положение: женат/замужем 12339
Семейное положение: гражданский брак 4151
Семейное положение: вдова/вдовец 959
Семейное положение: в разводе 1195
Семейное положение: не женат/не замужем 2810

Цель кредита: недвижимость 10811
Цель кредита: автомобиль 4306
Цель кредита: образование 4013
Цель кредита: свадьба 2324


In [29]:
# данные по количеству должников с детьми
print('Бездетных должников:',len(data.loc[(data['debt'] == 1) & (data['kindergarten'] == 'бездетный')]))
print('Обычнодетных должников:',len(data.loc[(data['debt'] == 1) & (data['kindergarten'] == 'обычнодетный')]))
print('Многодетных должников:',len(data.loc[(data['debt'] == 1) & (data['kindergarten'] == 'многодетный')]))
print()

# данные по должникам и уровню дохода 
print('Должники с низким доходом:',len(data.loc[(data['debt'] == 1) & (data['richness'] == 'низкий доход')]))
print('Должники с доходом ниже среднего:',len(data.loc[(data['debt'] == 1) & (data['richness'] == 'ниже среднего')]))
print('Должники с доходом выше среднего:',len(data.loc[(data['debt'] == 1) & (data['richness'] == 'выше среднего')]))
print('Должники с высоким доходом:',len(data.loc[(data['debt'] == 1) & (data['richness'] == 'высокий')]))
print()   

# данные по должникам и семейному статусу
print('Женатых/замужних должников:',len(data.loc[(data['debt'] == 1) & (data['family_status_id'] == 0)]))
print('Должников в гражданском браке:',len(data.loc[(data['debt'] == 1) & (data['family_status_id'] == 1)]))
print('Должников вдов и вдовцов:',len(data.loc[(data['debt'] == 1) & (data['family_status_id'] == 2)]))
print('Разведённых должников:',len(data.loc[(data['debt'] == 1) & (data['family_status_id'] == 3)]))
print('Неженатых/незамужних должников:',len(data.loc[(data['debt'] == 1) & (data['family_status_id'] == 4)]))
print()

# данные по должникам и целям кредита
print('Кредит на недвижимость:',len(data.loc[(data['debt'] == 1) & (data['purpose'] == 'недвижимость')]))
print('Кредит на автомобиль:',len(data.loc[(data['debt'] == 1) & (data['purpose'] == 'автомобиль')]))
print('Кредит на обраование:',len(data.loc[(data['debt'] == 1) & (data['purpose'] == 'образование')]))
print('Кредит на свадьбу:',len(data.loc[(data['debt'] == 1) & (data['purpose'] == 'свадьба')]))

Бездетных должников: 1071
Обычнодетных должников: 639
Многодетных должников: 31

Должники с низким доходом: 427
Должники с доходом ниже среднего: 473
Должники с доходом выше среднего: 458
Должники с высоким доходом: 383

Женатых/замужних должников: 931
Должников в гражданском браке: 388
Должников вдов и вдовцов: 63
Разведённых должников: 85
Неженатых/незамужних должников: 274

Кредит на недвижимость 782
Кредит на автомобиль: 403
Кредит на обраование: 370
Кредит на свадьбу: 186


In [36]:
data_1 = [
    ['Бездетных должников', 14167, 1071],
    ['Обычнодетных должников', 6907, 639],
    ['Многодетных должников', 380, 31],
    ['Должники с низким доходом', 5364, 427],
    ['Должники с доходом ниже среднег', 5363, 473],
    ['Должники с доходом выше среднего',5363, 458],
    ['Должники с высоким доходом', 5364, 383],
    ['Женатых/замужних должников', 12339, 931],
    ['Должников в гражданском браке', 4151, 388],
    ['Должников вдов и вдовцов', 959, 63],
    ['Разведённых должников', 1195, 85],
    ['Неженатых/незамужних должников', 2810, 274],
    ['Кредит на недвижимость', 10811, 782],
    ['Кредит на автомобиль', 4306, 403],
    ['Кредит на обраование', 4013, 370],
    ['Кредит на свадьбу', 2324, 186]
]

columns = ['category_clients', 'totaly_clients', 'debt_clients']

data_2 = pd.DataFrame(data = data_1, columns = columns)

data_2['part_debt_%'] = data_2['debt_clients'] / data_2['totaly_clients'] * 100

display(data_2)

category_clients  totaly_clients  debt_clients  \
0                Бездетных должников           14167          1071   
1             Обычнодетных должников            6907           639   
2              Многодетных должников             380            31   
3          Должники с низким доходом            5364           427   
4    Должники с доходом ниже среднег            5363           473   
5   Должники с доходом выше среднего            5363           458   
6         Должники с высоким доходом            5364           383   
7         Женатых/замужних должников           12339           931   
8      Должников в гражданском браке            4151           388   
9           Должников вдов и вдовцов             959            63   
10             Разведённых должников            1195            85   
11    Неженатых/незамужних должников            2810           274   
12            Кредит на недвижимость           10811           782   
13              Кредит на автомобиль            4306           403   
14              Кредит на обраование            4013           370   
15                 Кредит на свадьбу            2324           186   

    part_debt_%  
0      7.559822  
1      9.251484  
2      8.157895  
3      7.960477  
4      8.819690  
5      8.539996  
6      7.140194  
7      7.545182  
8      9.347145  
9      6.569343  
10     7.112971  
11     9.750890  
12     7.233373  
13     9.359034  
14     9.220035  
15     8.003442

<div class="alert alert-danger">
<b>Комментарий ревьюера: </b> 
    
Ты предложил интересный способ проверки задолжности, однако он не совсем точный и очень сложно воспринимается. В идеале на данном этапе нужно составить табличку следующего вида: для каждой категории посчитать количество клиентов в категории, количество должников в категории и процент должников (второй столбец делим на первый :) ). Это самый быстрый способ, который учитывает и процент задолжности в зависимости от категории и то, насколько мы можем доверять этому проценту. Если выборка достаточно большая, то процент репрезентативен, если это 2 человека из 10, то, конечно, по этой категории выводы лучше не делать. </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Предложенный вариант рассчёта дал ничего не говорящие примерно одинаковые проценты для каждой категории.
    
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2: </b>

В банковской сфере даже 1 процент может быть значим. В одну табличку можно было бы не объединять, тогда мы бы просто посмотрели для каждого параметра какая категория чаще всего не выплачивает кредит в срок. Не всегда нужно стремиться к тому, чтобы была огромная разница в процентах. Сейчас проценты посчитаны верно, быстрее всего это можно было сделать вот так:
    
```python
table = df.pivot_table(index = 'purpose', 
                                          values = 'debt', 
                                          aggfunc = ['count','sum'])

table['ratio'] = (table['sum'] / table['count'] * 100)
    
```
    
Твой метод не является неправильным! Вообще в реальной практике зависимость задолжности от параметров смотрят, используя статистические методы, мы же смотрим и оцениваем тенденции. И на практике хотят мы встречались с тем, что от нас хотели видеть именно такой процент.
    
    
Так что я считаю, что эти два метода отлично дополняют друг друга.
</div>

## Шаг 4. Общий вывод

1. В исходной таблице присутствует много артефактов, аномальных значений, пропущенных значений.

2. Пропущенные значения в колонках "трудовой стаж" и "ежемесячный заработок", вероятно не случаен, а объясняется сокрытием данных клиентом, или невозможностью предоставить подтверждающих "трудовой стаж" и "ежемесячный заработок" документов, что возможно в случае, например, неофициальной работы.

3. Часть аномалий не поддаётся приведению в нормальный вид, удовлетворяющий физическому. биологическому и здравому смыслу - таковы значения трудового стажа в большой части данных.

4. Часть аномальных значений поддаются обработке, и, вероятно объясняются ошибкой оператора. Минус один ребёнок в столбце "дети", вероятно является записью "тире" один ребёнок. Аномально большое число в столбце "дети" - 20 детей, вероятно объясняется небрежностью внесения данных - клавиши "0" и "2" расположены рядом на клавиатуре.

5. Таблица содержит данные о клиентах, имевших задолженность по возврату кредитов. Всего в таблице  1741 запись, что соответствует такому же количеству человек.

Из 1741 человека, имевших задолженность по возврату кредитов:

- бездетные клиенты составляют: 61.52%
- клиенты имеющие 1 или 2 детей, составляют: 36.70%
- многодетные клиенты, т.е., имеющие от 3 до 5 детей, составляют: 1.78%.


- женатые и замужние: 53.48%
- состоящие в гражданском браке: 22.29%
- вдовцы и вдовы: 3.62%
- состоят в разводе: 4.88%
- незамужние и неженатые: 15.74%


- малоимущих (имеющих заработок до 80 000 рублей в месяц): 9.99%
- со средним заработком (имеющих заработок от 80 000 до 180 000 рублей в месяц): 62.55%
- обеспеченных (имеющих заработок свыше 180 000 рублей в месяц): 27.46%


- имеют цель кредита: недвижимость: 44.92%
- имеют цель кредита: автомобиль: 23.15%
- имеют цель кредита: получение образования: 21.25%
- имеют цель кредита: проведение свадьбы: 10.68%


Таким образом, по результатам исследования, наиболее вероятно кредит вернёт в срок многодетный малоимущий вдовец (вдова), берущий кредит на проведение свадебного торжества. Наиболее вероятно кредит не вернёт в срок бездетный замужний/женатый клиент со средним заработком (от 80 000 до 180 000 рублей в месяц) с кредитом на недвижимость.



<div class="alert alert-success">
<b>Комментарий ревьюера: </b> 

Классно, что в выводе ты подробно расписал все про проценты задолжности. Это то, что и требовалось в цели проекта. Последний абзай как раз обобщает полученные результаты.</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Проект я исправил в соответствии с предложенными правками. Теперь своим проектом эту работу я не считаю. Получившимся результатом я сильно огорчён и расстроен. 
    
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2: </b>

Андрей, цель ревью ни в коем случае не растроить, а проверить на наличие ошибок и дать советы по усовершенствованию! Я нигде не сказала, что ты был неправ, я лишь предложила варианты, которые, по моему опыту, точнее. В первой версии ты не совершил критических ошибок, так что если ты после приведенных мною аргументов все еще не согласен с предложенными идеями, то ты, конечно, можешь оставить свою первую версию проекта. Для аналитика важный навык - умение отстаивать свое видение решения задачи. 
    
Ты проделал огромную работу и сделал очень солидный проект, ты точно можешь собой гордиться! И я уверена, что те навыки, которые ты получил, пока добавлял код по моим рекомендациям, будут полезны!
    
Я тоже немного растроена, что у тебя остался такой негативный осадок после нашего взаимодействия, надеюсь, что со следующим проектом такого не будет. Главное, помни - ты всегда можешь настоять на своем решение. Практикум - это не совковая система, студент и ревьюер находятся в диалоге, а не в иерархии.
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [х ]  файл изучен;
- [х ]  определены пропущенные значения;
- [х ]  заполнены пропущенные значения;
- [х ]  есть пояснение, какие пропущенные значения обнаружены;
- [х ]  описаны возможные причины появления пропусков в данных;
- [х ]  объяснено, по какому принципу заполнены пропуски;
- [х ]  заменен вещественный тип данных на целочисленный;
- [ х]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [х ]  удалены дубликаты;
- [ х]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [ х]  описаны возможные причины появления дубликатов в данных;
- [х ]  выделены леммы в значениях столбца с целями получения кредита;
- [ х]  описан процесс лемматизации;
- [х ]  данные категоризированы;
- [х ]  есть объяснение принципа категоризации данных;
- [х ]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [ х]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [х ]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [х ]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [х ]  в каждом этапе есть выводы;
- [х ]  есть общий вывод.

<div class="alert alert-success">
<b>Комментарий ревьюера: </b> 

Андрей, ты хорошо справился со своим первым полноценным проектом: тщательно подготовил данные для анализа, провел лемматизацию и категоризацию, подробнейшим образом все аргументировал

Предлагаю тебе немного доработать проект, чтобы довести его до совершенства:

1. Доработать категоризацию по уровню дохода.
2. Для каждой категории посчитать количество клиентов, количество должников и процент должников.

Если будут вопросы, обращайся, с удовольствием на них отвечу :)</div>